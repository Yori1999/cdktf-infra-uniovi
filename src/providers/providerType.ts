export enum ProviderType {
  DOCKER = "docker",
  AWS = "aws",
}
