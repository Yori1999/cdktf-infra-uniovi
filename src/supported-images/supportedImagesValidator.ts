export function validateImageExists(imageIdentifier: string): void {
  if (!imageIdentifier || imageIdentifier.trim().length === 0) {
    throw new Error(
      "This version of the machine doesn't exist for the given provider",
    );
  }
}
