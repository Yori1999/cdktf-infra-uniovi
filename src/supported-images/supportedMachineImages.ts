export enum AlpineVersion {
  LATEST = "latest",
}
export const supportedAlpineImages = Object.freeze({
  [ProviderType.Docker]: {
    [AlpineVersion.LATEST]: "alpine:latest",
  },
  [ProviderType.AWS]: {
    [AlpineVersion.LATEST]: "ami-0d2e72039b74fb9d7", // Alpine Linux on AWS from Solve DevOps, from Marketplace
  },
}) satisfies Record<ProviderType, Record<AlpineVersion, string>>;

export enum DebianVersion {
  LATEST = "latest",
  DEBIAN_12 = "debian12",
  DEBIAN_11 = "debian11",
}
export const supportedDebianImages = Object.freeze({
  [ProviderType.Docker]: {
    [DebianVersion.LATEST]: "debian:latest",
    [DebianVersion.DEBIAN_12]: "debian:bookworm",
    [DebianVersion.DEBIAN_11]: "debian:bullseye",
  },
  [ProviderType.AWS]: {
    [DebianVersion.LATEST]: "ami-0fbb72557598f5284", // all AMIs are from AWS Marketplace
    [DebianVersion.DEBIAN_12]: "ami-0fbb72557598f5284", // Debian 12 (HVM), EBS General Purpose (SSD) Volume Type. Community developed free GNU/Linux distribution.
    [DebianVersion.DEBIAN_11]: "ami-08b11b3defb3d4b52", // Debian 11
  },
}) satisfies Record<ProviderType, Record<DebianVersion, string>>;

export enum UbuntuVersion {
  LATEST = "latest",
  UBUNTU_18 = "ubuntu18",
  UBUNTU_20 = "ubuntu20",
  UBUNTU_22 = "ubuntu22",
  UBUNTU_24 = "ubuntu24",
}
export const supportedUbuntuImages = Object.freeze({
  [ProviderType.Docker]: {
    [UbuntuVersion.LATEST]: "ubuntu:latest",
    [UbuntuVersion.UBUNTU_18]: "ubuntu:18.04",
    [UbuntuVersion.UBUNTU_20]: "ubuntu:20.04",
    [UbuntuVersion.UBUNTU_22]: "ubuntu:22.04",
    [UbuntuVersion.UBUNTU_24]: "ubuntu:24.04",
  },
  [ProviderType.AWS]: {
    [UbuntuVersion.LATEST]: "ami-01f23391a59163da9",
    [UbuntuVersion.UBUNTU_18]: "",
    [UbuntuVersion.UBUNTU_20]: "",
    [UbuntuVersion.UBUNTU_22]: "ami-021d9f8e43481e7da", // Ubuntu Server 22.04 LTS (HVM), SSD Volume Type
    [UbuntuVersion.UBUNTU_24]: "ami-01f23391a59163da9", // Ubuntu Server 24.04 LTS (HVM), SSD Volume Type
  },
}) satisfies Record<ProviderType, Record<UbuntuVersion, string>>;
// satisfies preserva claves exactas sin ampliarlas, a diferencia de : Type, que las "ensancha".
