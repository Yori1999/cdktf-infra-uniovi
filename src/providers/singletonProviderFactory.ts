import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { DockerProvider } from "@cdktf/provider-docker/lib/provider";
import { TerraformProvider } from "cdktf";
import { Construct } from "constructs";
import { ProviderType } from "./providerType";

export class SingletonProviderFactory {
  // We could've chosen not to return anything but... we might need it
  static getProvider(
    providerType: ProviderType,
    scope: Construct,
    existingProvider?: TerraformProvider,
  ): TerraformProvider {
    if (existingProvider) {
      return existingProvider;
    }

    if (!this.providers[providerType]) {
      const provider = SingletonProviderFactory.createProvider(
        providerType,
        scope,
      );
      this.providers[providerType] = provider;
    }

    return this.providers[providerType];
  }

  private static providers: { [key in ProviderType]?: TerraformProvider } = {};

  /**
   * Creates a provider based on a given type, with either the default properties or the properties passed by the user
   * @param providerType the type of the provider to create
   * @param scope
   * @param properties a set of properties that may override the default ones
   * @returns
   */
  private static createProvider(
    providerType: ProviderType,
    scope: Construct,
  ): TerraformProvider {
    switch (providerType) {
      case ProviderType.DOCKER:
        return new DockerProvider(scope, "docker", {});
      case ProviderType.AWS:
        return new AwsProvider(scope, "aws", {
          region: "eu-west-1",
        });
    }
  }

  // Prevents instantiation, just in case
  private constructor() {}
}
