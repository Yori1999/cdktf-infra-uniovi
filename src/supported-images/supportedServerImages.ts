import { DebianVersion, supportedDebianImages } from "./supportedMachineImages";
import { ProviderType } from "../providers/providerType";

export enum NginxVersion {
  LATEST = "latest",
}
export const supportedNginxImages = Object.freeze({
  [ProviderType.DOCKER]: {
    [NginxVersion.LATEST]: "nginx:latest",
  },
  [ProviderType.AWS]: {
    [NginxVersion.LATEST]:
      supportedDebianImages[ProviderType.AWS][DebianVersion.LATEST],
  },
}) satisfies Record<ProviderType, Record<NginxVersion, string>>;
export const supportedHardenedNginxImages = Object.freeze({
  [ProviderType.DOCKER]: {
    [NginxVersion.LATEST]: "rapidfort/nginx-official:latest",
  },
  [ProviderType.AWS]: {
    [NginxVersion.LATEST]: "ami-04ca7bd38fd01789c", // NGINX Plus App Protect - Ubuntu 20.04 Developer AMI
  },
}) satisfies Record<ProviderType, Record<NginxVersion, string>>;

export enum ApacheVersion {
  LATEST = "latest",
  APACHE_DEBIAN = "apacheDebian",
  APACHE_ALPINE = "apacheAlpine",
}
export const supportedApacheImages = Object.freeze({
  [ProviderType.DOCKER]: {
    [ApacheVersion.LATEST]: "httpd:latest",
    [ApacheVersion.APACHE_DEBIAN]: "httpd:bookworm",
    [ApacheVersion.APACHE_ALPINE]: "httpd:alpine",
  },
  [ProviderType.AWS]: {
    [ApacheVersion.LATEST]:
      supportedDebianImages[ProviderType.AWS][DebianVersion.LATEST],
    [ApacheVersion.APACHE_DEBIAN]:
      supportedDebianImages[ProviderType.AWS][DebianVersion.LATEST],
    [ApacheVersion.APACHE_ALPINE]: "", // doesn't exist, so we cannot assign this version but we need to include it
  },
}) satisfies Record<ProviderType, Record<ApacheVersion, string>>;
export const supportedHardenedApacheImages = Object.freeze({
  [ProviderType.DOCKER]: {
    [ApacheVersion.LATEST]: "rapidfort/apache-official:latest", // It is equivalent to httpd:latest
    [ApacheVersion.APACHE_DEBIAN]: "rapidfort/apache-official:2.4.63-bookworm",
    [ApacheVersion.APACHE_ALPINE]:
      "rapidfort/apache-official:2.4.63-alpine3.22",
  },
  [ProviderType.AWS]: {
    [ApacheVersion.LATEST]:
      supportedDebianImages[ProviderType.AWS][DebianVersion.LATEST],
    [ApacheVersion.APACHE_DEBIAN]:
      supportedDebianImages[ProviderType.AWS][DebianVersion.LATEST],
    [ApacheVersion.APACHE_ALPINE]: "", // doesn't exist, so we cannot assign this version but we need to include it
  },
}) satisfies Record<ProviderType, Record<ApacheVersion, string>>;
