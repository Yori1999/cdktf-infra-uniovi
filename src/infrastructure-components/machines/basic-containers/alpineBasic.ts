import { TerraformProvider } from "cdktf";
import { Construct } from "constructs";
import {
  BasicMachineComponentPropsInterface,
  InternalMachineComponentPropsInterface,
} from "../../../props/props";
import { IDeployStrategy } from "../../../providers/providerDeployStrategy/deployStrategy";
import { ProviderDeployStrategyFactory } from "../../../providers/providerDeployStrategy/providerDeployStrategyFactory";
import { SingletonProviderFactory } from "../../../providers/singletonProviderFactory";
import {
  AlpineVersion,
  supportedAlpineImages,
} from "../../../supported-images/supportedMachineImages";
import { checkIsValidId } from "../../../utils/stringUtils";

export class AlpineBasic extends Construct {
  constructor(
    scope: Construct,
    id: string,
    version: AlpineVersion,
    machineProps: BasicMachineComponentPropsInterface,
    provider?: TerraformProvider,
  ) {
    // necesario normalizar el id
    checkIsValidId(id);

    super(scope, id);

    const deployWith: ProviderType = machineProps.providerType;

    if (Object.keys(supportedAlpineImages[deployWith]).length > 0) {
      const imageWithTag = (
        supportedAlpineImages[deployWith] as Record<AlpineVersion, string>
      )[version];
      if (imageWithTag != undefined) {
        SingletonProviderFactory.getProvider(deployWith, this, provider);
        const deployStrategy: IDeployStrategy =
          ProviderDeployStrategyFactory.getProviderDeployStrategy(deployWith);
        deployStrategy.deployBasicMachine(
          this,
          id,
          machineProps,
          this.getAdditionalProps(deployWith, imageWithTag),
        );
      }
    }
  }

  protected getAdditionalProps(
    providerType: ProviderType,
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface {
    switch (providerType) {
      case ProviderType.Docker:
        return {
          dockerProps: {
            imageName: imageIdentifier,
          },
        };
      case ProviderType.AWS:
        return {
          awsProps: {
            ami: imageIdentifier,
          },
        };
      default:
        return {};
    }
  }
}
