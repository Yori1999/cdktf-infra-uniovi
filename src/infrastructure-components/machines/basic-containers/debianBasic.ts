import { TerraformProvider } from "cdktf";
import { Construct } from "constructs";
import {
  BasicMachineComponentPropsInterface,
  InternalMachineComponentPropsInterface,
} from "../../../props/props";
import { IDeployStrategy } from "../../../providers/providerDeployStrategy/deployStrategy";
import { ProviderDeployStrategyFactory } from "../../../providers/providerDeployStrategy/providerDeployStrategyFactory";
import { ProviderType } from "../../../providers/providerType";
import { SingletonProviderFactory } from "../../../providers/singletonProviderFactory";
import { validateImageExists } from "../../../supported-images/supportedImagesValidator";
import {
  DebianVersion,
  supportedDebianImages,
} from "../../../supported-images/supportedMachineImages";
import { checkIsValidId } from "../../../utils/stringUtils";

export class DebianBasic extends Construct {
  protected createdDebianMachine?: Construct;

  constructor(
    scope: Construct,
    id: string,
    version: DebianVersion,
    machineProps: BasicMachineComponentPropsInterface,
    provider?: TerraformProvider,
  ) {
    // necesario normalizar el id
    checkIsValidId(id);

    super(scope, id);

    const deployWith: ProviderType = machineProps.providerType;
    const imageIdentifier: string = supportedDebianImages[deployWith][version];

    validateImageExists(imageIdentifier);

    SingletonProviderFactory.getProvider(deployWith, this, provider);
    const deployStrategy: IDeployStrategy =
      ProviderDeployStrategyFactory.getProviderDeployStrategy(deployWith);
    this.createdDebianMachine = deployStrategy.deployBasicMachine(
      this,
      id,
      machineProps,
      this.getAdditionalProps(deployWith, imageIdentifier),
    );
  }

  protected getAdditionalProps(
    providerType: ProviderType,
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface {
    switch (providerType) {
      case ProviderType.DOCKER:
        return {
          dockerProps: {
            imageName: imageIdentifier,
          },
        };
      case ProviderType.AWS:
        return {
          awsProps: {
            ami: imageIdentifier,
          },
        };
      default:
        return {};
    }
  }
}
