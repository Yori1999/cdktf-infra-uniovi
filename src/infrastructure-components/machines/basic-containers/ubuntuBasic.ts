import { TerraformProvider } from "cdktf";
import { Construct } from "constructs";
import {
  BasicMachineComponentPropsInterface,
  InternalMachineComponentPropsInterface,
} from "../../../props/props";
import { IDeployStrategy } from "../../../providers/providerDeployStrategy/deployStrategy";
import { ProviderDeployStrategyFactory } from "../../../providers/providerDeployStrategy/providerDeployStrategyFactory";
import { ProviderType } from "../../../providers/providerType";
import { SingletonProviderFactory } from "../../../providers/singletonProviderFactory";
import { validateImageExists } from "../../../supported-images/supportedImagesValidator";
import {
  supportedUbuntuImages,
  UbuntuVersion,
} from "../../../supported-images/supportedMachineImages";
import { checkIsValidId } from "../../../utils/stringUtils";

export class UbuntuBasic extends Construct {
  constructor(
    scope: Construct,
    id: string,
    version: UbuntuVersion,
    machineProps: BasicMachineComponentPropsInterface,
    provider?: TerraformProvider,
  ) {
    // necesario normalizar el id
    checkIsValidId(id);

    super(scope, id);

    const deployWith: ProviderType = machineProps.providerType;
    const imageIdentifier: string = supportedUbuntuImages[deployWith][version];

    validateImageExists(imageIdentifier);

    SingletonProviderFactory.getProvider(deployWith, this, provider);
    const deployStrategy: IDeployStrategy =
      ProviderDeployStrategyFactory.getProviderDeployStrategy(deployWith);
    deployStrategy.deployBasicMachine(
      this,
      id,
      machineProps,
      this.getAdditionalProps(deployWith, imageIdentifier),
    );
  }

  protected deploy(
    strategy: IDeployStrategy,
    id: string,
    machineProps: BasicMachineComponentPropsInterface,
    imageIdentifier: string,
  ): Construct {
    return strategy.deployBasicMachine(
      this,
      id,
      machineProps,
      this.getAdditionalProps(machineProps.providerType, imageIdentifier),
    );
  }

  protected getAdditionalProps(
    providerType: ProviderType,
    imageIdentifier: string,
  ): InternalMachineComponentPropsInterface {
    switch (providerType) {
      case ProviderType.DOCKER:
        return {
          dockerProps: {
            imageName: imageIdentifier,
          },
        };
      case ProviderType.AWS:
        return {
          awsProps: {
            ami: imageIdentifier,
          },
        };
      default:
        return {};
    }
  }
}
