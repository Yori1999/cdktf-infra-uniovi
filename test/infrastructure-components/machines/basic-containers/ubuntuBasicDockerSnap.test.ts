import { App, TerraformStack, Testing } from "cdktf";
import { UbuntuBasic } from "../../../../src/infrastructure-components/machines/basic-containers/ubuntuBasic";
import { ProviderType } from "../../../../src/providers/providerType";
import { UbuntuVersion } from "../../../../src/supported-images/supportedMachineImages";

function createTestStack(
  version: UbuntuVersion,
  useVolume: boolean,
): TerraformStack {
  const app = new App();
  const stack = new TerraformStack(app, `test-${version}`);

  new UbuntuBasic(stack, "Test", version, {
    providerType: ProviderType.DOCKER,
    dockerProps: {
      useVolume: useVolume,
    },
  });

  return stack;
}

const ubuntuVersions = [
  { version: UbuntuVersion.UBUNTU_18, expectedImage: "ubuntu:18.04" },
  { version: UbuntuVersion.UBUNTU_20, expectedImage: "ubuntu:20.04" },
  { version: UbuntuVersion.UBUNTU_22, expectedImage: "ubuntu:22.04" },
  { version: UbuntuVersion.UBUNTU_24, expectedImage: "ubuntu:24.04" },
  { version: UbuntuVersion.LATEST, expectedImage: "ubuntu:latest" },
];

describe("UbuntuBasic Docker container creation", () => {
  test.each(ubuntuVersions)(
    "creates a Docker container for %o",
    ({ version, expectedImage }) => {
      const stack = createTestStack(version, false);
      const synth = Testing.synth(stack);
      const parsed = JSON.parse(synth);

      expect(parsed).toMatchSnapshot();
      expect(parsed.resource.docker_container).toBeDefined();

      const image = parsed.resource.docker_image["Test_Test-image_5197E2DB"];
      expect(image.name).toBe(expectedImage);
    },
  );

  test("creates a Docker container with volume", () => {
    const stack = createTestStack(UbuntuVersion.UBUNTU_22, true);
    const synth = Testing.synth(stack);
    const parsed = JSON.parse(synth);

    expect(parsed).toMatchSnapshot();

    const container =
      parsed.resource.docker_container["Test_Test-container_72A6054C"];
    expect(container.volumes).toHaveLength(1);
  });
});
