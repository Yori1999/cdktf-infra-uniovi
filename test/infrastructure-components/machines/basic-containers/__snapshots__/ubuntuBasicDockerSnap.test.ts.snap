// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UbuntuBasic Docker container creation creates a Docker container for { version: 'latest', expectedImage: 'ubuntu:latest' } 1`] = `
{
  "output": {
    "Test_Test-public-ip_348420EB": {
      "value": "\${docker_container.Test_Test-container_72A6054C.network_data[0].ip_address}",
    },
  },
  "resource": {
    "docker_container": {
      "Test_Test-container_72A6054C": {
        "command": [
          "sleep",
          "infinity",
        ],
        "image": "\${docker_image.Test_Test-image_5197E2DB.name}",
        "name": "Test-container",
      },
    },
    "docker_image": {
      "Test_Test-image_5197E2DB": {
        "keep_locally": false,
        "name": "ubuntu:latest",
      },
    },
  },
}
`;

exports[`UbuntuBasic Docker container creation creates a Docker container for { version: 'ubuntu18', expectedImage: 'ubuntu:18.04' } 1`] = `
{
  "output": {
    "Test_Test-public-ip_348420EB": {
      "value": "\${docker_container.Test_Test-container_72A6054C.network_data[0].ip_address}",
    },
  },
  "provider": {
    "docker": [
      {},
    ],
  },
  "resource": {
    "docker_container": {
      "Test_Test-container_72A6054C": {
        "command": [
          "sleep",
          "infinity",
        ],
        "image": "\${docker_image.Test_Test-image_5197E2DB.name}",
        "name": "Test-container",
      },
    },
    "docker_image": {
      "Test_Test-image_5197E2DB": {
        "keep_locally": false,
        "name": "ubuntu:18.04",
      },
    },
  },
  "terraform": {
    "required_providers": {
      "docker": {
        "source": "kreuzwerker/docker",
        "version": "3.0.2",
      },
    },
  },
}
`;

exports[`UbuntuBasic Docker container creation creates a Docker container for { version: 'ubuntu20', expectedImage: 'ubuntu:20.04' } 1`] = `
{
  "output": {
    "Test_Test-public-ip_348420EB": {
      "value": "\${docker_container.Test_Test-container_72A6054C.network_data[0].ip_address}",
    },
  },
  "resource": {
    "docker_container": {
      "Test_Test-container_72A6054C": {
        "command": [
          "sleep",
          "infinity",
        ],
        "image": "\${docker_image.Test_Test-image_5197E2DB.name}",
        "name": "Test-container",
      },
    },
    "docker_image": {
      "Test_Test-image_5197E2DB": {
        "keep_locally": false,
        "name": "ubuntu:20.04",
      },
    },
  },
}
`;

exports[`UbuntuBasic Docker container creation creates a Docker container for { version: 'ubuntu22', expectedImage: 'ubuntu:22.04' } 1`] = `
{
  "output": {
    "Test_Test-public-ip_348420EB": {
      "value": "\${docker_container.Test_Test-container_72A6054C.network_data[0].ip_address}",
    },
  },
  "resource": {
    "docker_container": {
      "Test_Test-container_72A6054C": {
        "command": [
          "sleep",
          "infinity",
        ],
        "image": "\${docker_image.Test_Test-image_5197E2DB.name}",
        "name": "Test-container",
      },
    },
    "docker_image": {
      "Test_Test-image_5197E2DB": {
        "keep_locally": false,
        "name": "ubuntu:22.04",
      },
    },
  },
}
`;

exports[`UbuntuBasic Docker container creation creates a Docker container for { version: 'ubuntu24', expectedImage: 'ubuntu:24.04' } 1`] = `
{
  "output": {
    "Test_Test-public-ip_348420EB": {
      "value": "\${docker_container.Test_Test-container_72A6054C.network_data[0].ip_address}",
    },
  },
  "resource": {
    "docker_container": {
      "Test_Test-container_72A6054C": {
        "command": [
          "sleep",
          "infinity",
        ],
        "image": "\${docker_image.Test_Test-image_5197E2DB.name}",
        "name": "Test-container",
      },
    },
    "docker_image": {
      "Test_Test-image_5197E2DB": {
        "keep_locally": false,
        "name": "ubuntu:24.04",
      },
    },
  },
}
`;

exports[`UbuntuBasic Docker container creation creates a Docker container with volume 1`] = `
{
  "output": {
    "Test_Test-public-ip_348420EB": {
      "value": "\${docker_container.Test_Test-container_72A6054C.network_data[0].ip_address}",
    },
  },
  "resource": {
    "docker_container": {
      "Test_Test-container_72A6054C": {
        "command": [
          "sleep",
          "infinity",
        ],
        "image": "\${docker_image.Test_Test-image_5197E2DB.name}",
        "name": "Test-container",
        "volumes": [
          {
            "container_path": "/mnt/data",
            "volume_name": "\${docker_volume.Test_Test-volume_F1ABD0A9.name}",
          },
        ],
      },
    },
    "docker_image": {
      "Test_Test-image_5197E2DB": {
        "keep_locally": false,
        "name": "ubuntu:22.04",
      },
    },
    "docker_volume": {
      "Test_Test-volume_F1ABD0A9": {
        "name": "Test-data",
      },
    },
  },
}
`;
